doc/man/man1/sshare.1
.TH sshare "1" "Slurm Commands" "Slurm 26.11" "Slurm Commands"

.SH "NAME"
sshare \- Tool for listing the shares of associations to a cluster.

.SH "SYNOPSIS"
\fBsshare\fR [\fIOPTIONS\fR...]

.SH "DESCRIPTION"
\fBsshare\fR is used to view Slurm share information. This command is
only viable when running with the priority/multifactor plugin.
The sshare information is derived from a database with the interface
being provided by \fBslurmdbd\fR (Slurm Database daemon) which is
read in from the slurmctld and used to process the shares available
to a given association. sshare provides Slurm share information of
Account, User, Raw Shares, Normalized Shares, Raw Usage, Normalized
Usage, Effective Usage, the Fair\-share factor, the GrpTRESMins limit,
Partitions and accumulated currently running TRES\-minutes for each association.


.SH "OPTIONS"

.TP
\fB\-A\fR, \fB\-\-accounts\fR=<\fIaccount\fR>
Display information for specific accounts (comma separated list).
.IP

.TP
\fB\-a\fR, \fB\-\-all\fR
Display information for all users.
.IP

.TP
\fB\-M\fR, \fB\-\-clusters\fR=<\fIstring\fR>
Clusters to issue commands to.
Note that the \fBslurmdbd\fR must be up for this option to work properly, unless
running in a federation with \fBFederationParameters=fed_display\fR configured.
.IP

.TP
\fB\-o\fR, \fB\-\-format\fR=<\fIoutput_format\fR>
Comma separated list of fields (use
"\-\-helpformat" for a list of available fields).
.IP

.TP
\fB\-\-help\fR
Display a description of sshare options and commands.
.IP

.TP
\fB\-l\fR, \fB\-\-long\fR
Long listing \- includes the normalized usage information.
.IP

.TP
\f3\-\-json\fP, \f3\-\-json\fP=\fIlist\fR, \f3\-\-json\fP=<\fIdata_parser\fR>
Dump information as JSON using the default data_parser plugin or explicit
data_parser with parameters. Sorting and formatting arguments will be ignored.
.IP

.TP
\fB\-n\fR, \fB\-\-noheader\fR
No header will be added to the beginning of the output.
.IP

.TP
\fB\-p\fR, \fB\-\-parsable\fR
Output will be '|' delimited with a '|' at the end.
.IP

.TP
\fB\-P\fR, \fB\-\-parsable2\fR
Output will be '|' delimited without a '|' at the end.
.IP

.TP
\fB\-m\fR, \fB\-\-partition\fR
If there are association based partitions in the system
print their names.
.IP

.TP
\fB\-\-usage\fR
Display a description of sshare options and commands.
.IP

.TP
\fB\-u\fR, \fB\-\-users\fR=<\fIuser_list\fR>
Display information for specific users (comma separated list).
.IP

.TP
\fB\-U\fR, \fB\-\-Users\fR
If specified only the users information are printed, the parent
and ancestors are not displayed.
.IP

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Display more information about the specified options.
.IP

.TP
\fB\-V\fR, \fB\-\-version\fR
Display the version number of sshare.
.IP

.TP
\f3\-\-yaml\fP, \f3\-\-yaml\fP=\fIlist\fR, \f3\-\-yaml\fP=<\fIdata_parser\fR>
Dump information as YAML using the default data_parser plugin or explicit
data_parser with parameters. Sorting and formatting arguments will be ignored.
.IP

.SH "SSHARE OUTPUT FIELDS"

.TP
\f3Account\fP
The Account.
.IP

.TP
\f3User\fP
The User.
.IP

.TP
\f3Raw Shares\fP
The raw shares assigned to the user or account.
.IP

.TP
\f3Norm Shares\fP
The shares assigned to the user or account normalized to the total
number of assigned shares.
.IP

.TP
\f3Raw Usage\fP
The number of tres\-seconds (cpu\-seconds if TRESBillingWeights is not defined)
of all the jobs charged to the account or user. This number will decay over
time when PriorityDecayHalfLife is defined.
.IP

.TP
\f3Norm Usage\fP (only appears with \fBsshare \-l\fR option)
The Raw Usage normalized to the total number of tres\-seconds of all
jobs run on the cluster, subject to the PriorityDecayHalfLife decay
when defined.
.IP

.TP
\f3Effectv Usage\fP
The Effective Usage augments the normalized usage to account for usage
from sibling accounts.
.IP

.TP
\f3FairShare\fP
The Fair\-Share factor, based on a user's assigned shares and
the effective usage charged to them.
.IP

.TP
\f3GrpTRESMins\fP
The TRES\-minutes limit set on the account. The total number of cpu
minutes that can possibly be used by past, present and future jobs
running from this account and its children.
.IP

.TP
\f3GrpTRESRaw\fP
The raw TRES usage that has been used by jobs running from
this account and its children.
.IP

.TP
\f3TRESRunMins\fP
The number of TRES\-minutes allocated by jobs currently running against
the account. Used to limit the combined total number of TRES minutes
used by all jobs running with this account and its children.
This takes into consideration time limit of running jobs and consumes it,
if the limit is reached no new jobs are started until other jobs finish
to allow time to free up.
.IP

.SH "FAIR_TREE MODIFICATIONS"
When PriorityFlags=FAIR_TREE is set (the default, unless NO_FAIR_TREE is set),
calculations are done differently.
As a result, the following fields are added or modified:

.TP
\f3Norm Shares\fP
The shares assigned to the user or account normalized to the total
number of assigned shares within the level.
.IP

.TP
\f3Effectv Usage\fP
Effectv Usage is the association's usage normalized with its parent.
.IP

.TP
\f3Level FS\fP (only appears with \fBsshare \-l\fR option)
This is the association's fairshare value compared to its siblings, calculated
as Norm Shares / Effectv Usage. If an association is over\-served, the value is
between 0 and 1. If an association is under\-served, the value is greater than 1.
Associations with no usage receive the highest possible value, infinity.
.IP

.LP
More information about Fair Tree can be found in doc/html/fair_tree.html or
at https://slurm.schedmd.com/fair_tree.html

.SH "PERFORMANCE"
.PP
Executing \fBsshare\fR sends a remote procedure call to \fBslurmctld\fR. If
enough calls from \fBsshare\fR or other Slurm client commands that send remote
procedure calls to the \fBslurmctld\fR daemon come in at once, it can result in
a degradation of performance of the \fBslurmctld\fR daemon, possibly resulting
in a denial of service.
.PP
Do not run \fBsshare\fR or other Slurm client commands that send remote
procedure calls to \fBslurmctld\fR from loops in shell scripts or other
programs. Ensure that programs limit calls to \fBsshare\fR to the minimum
necessary for the information you are trying to gather.

.SH "ENVIRONMENT VARIABLES"
.PP
Some \fBsshare\fR options may be set via environment variables. These
environment variables, along with their corresponding options, are listed below.
(Note: Command line options will always override these settings.)

.TP 20
\fBSLURM_CONF\fR
The location of the Slurm configuration file.
.IP

.TP
\fBSLURM_DEBUG_FLAGS\fR
Specify debug flags for sshare to use. See DebugFlags in the
\fBslurm.conf\fR(5) man page for a full list of flags. The environment
variable takes precedence over the setting in the slurm.conf.
.IP

.TP
\fBSLURM_JSON\fR
Control JSON serialization:
.IP
.RS
.TP
\fBcompact\fR
Output JSON as compact as possible.
.IP

.TP
\fBpretty\fR
Output JSON in pretty format to make it more readable.
.IP
.RE

.TP
\fBSLURM_YAML\fR
Control YAML serialization:
.IP
.RS
.TP
\fBcompact\fR Output YAML as compact as possible.
.IP

.TP
\fBpretty\fR Output YAML in pretty format to make it more readable.
.RE
.IP

.SH "EXAMPLES"

.TP
Display information about users in a particular account:
.IP
.nf
$ sshare \-A <Account>
.fi

.TP
Display information about a specific user in a parsable format:
.IP
.nf
$ sshare \-\-parsable \-\-users=<User>
.fi

.SH "COPYING"
Copyright (C) 2008 Lawrence Livermore National Security.
Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
.br
Copyright (C) 2010\-2022 SchedMD LLC.
.LP
This file is part of Slurm, a resource management program.
For details, see <https://slurm.schedmd.com/>.
.LP
Slurm is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your option)
any later version.
.LP
Slurm is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
details.

.SH "SEE ALSO"
\fBslurm.conf\fR(5),
\fBslurmdbd\fR(8)


NEWS
The contents of NEWS, from slurm-22.05 to the present, has been moved to
CHANGELOG.md and the CHANGELOG directory.


src/sshare/sshare.c
/*****************************************************************************\
 *  sshare.c -   tool for listing the shares of association in
 *               relationship to the cluster running on.
 *****************************************************************************
 *  Copyright (C) 2008 Lawrence Livermore National Security.
 *  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 *  Written by Danny Auble <da@llnl.gov>
 *  CODE-OCEC-09-009. All rights reserved.
 *
 *  This file is part of Slurm, a resource management program.
 *  For details, see <https://slurm.schedmd.com/>.
 *  Please also read the included file: DISCLAIMER.
 *
 *  Slurm is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the License, or (at your option)
 *  any later version.
 *
 *  In addition, as a special exception, the copyright holders give permission
 *  to link the code of portions of this program with the OpenSSL library under
 *  certain conditions as described in each individual source file, and
 *  distribute linked combinations including the two. You must obey the GNU
 *  General Public License in all respects for all of the code used other than
 *  OpenSSL. If you modify file(s) with this exception, you may extend this
 *  exception to your version of the file(s), but you are not obligated to do
 *  so. If you do not wish to do so, delete this exception statement from your
 *  version.  If you delete this exception statement from all source files in
 *  the program, then also delete it here.
 *
 *  Slurm is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 *  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 *  details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with Slurm; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA.
\*****************************************************************************/

#include <grp.h>

#include "src/common/data.h"
#include "src/common/proc_args.h"
#include "src/common/sercli.h"
#include "src/common/slurm_protocol_api.h"
#include "src/common/slurm_protocol_defs.h"
#include "src/common/uid.h"

#include "src/interfaces/data_parser.h"
#include "src/interfaces/priority.h"
#include "src/interfaces/serializer.h"

#include "src/sshare/sshare.h"

#define OPT_LONG_HELP  0x100
#define OPT_LONG_USAGE 0x101
#define OPT_LONG_AUTOCOMP 0x102
#define OPT_LONG_JSON 0x103
#define OPT_LONG_YAML 0x104

static int _addto_name_char_list(list_t *char_list, char *names, bool gid);
static int _single_cluster(int argc, char **argv,
			   shares_request_msg_t *req_msg);
static int _multi_cluster(int argc, char **argv, shares_request_msg_t *req_msg);
static char *   _convert_to_name(uint32_t id, bool is_gid);
static void     _print_version( void );
static void	_usage(void);
static void     _help_format_msg(void);

int exit_code;		/* sshare's exit code, =1 on any error at any time */
int quiet_flag;		/* quiet=1, verbose=-1, normal=0 */
int verbosity;		/* count of -v options */
uint32_t my_uid = 0;
list_t *clusters = NULL;
uint16_t options = 0;

int main (int argc, char **argv)
{
	int opt_char;
	log_options_t opts = LOG_OPTS_STDERR_ONLY;
	shares_request_msg_t req_msg;
	char *temp = NULL;
	int option_index;
	bool all_users = 0;

	static struct option long_options[] = {
		{"autocomplete", required_argument, 0, OPT_LONG_AUTOCOMP},
		{"accounts", 1, 0, 'A'},
		{"all",      0, 0, 'a'},
                {"helpformat",0,0, 'e'},
		{"long",     0, 0, 'l'},
		{"partition",0, 0, 'm'},
		{"cluster",  1, 0, 'M'},
		{"clusters", 1, 0, 'M'},
		{"noheader", 0, 0, 'n'},
		{"format",   1, 0, 'o'},
		{"parsable", 0, 0, 'p'},
		{"parsable2",0, 0, 'P'},
		{"users",    1, 0, 'u'},
		{"Users",    0, 0, 'U'},
		{"verbose",  0, 0, 'v'},
		{"version",  0, 0, 'V'},
		{"help",     0, 0, OPT_LONG_HELP},
		{"usage",    0, 0, OPT_LONG_USAGE},
		{"json",    optional_argument, 0, OPT_LONG_JSON},
		{"yaml",    optional_argument, 0, OPT_LONG_YAML},
		{NULL,       0, 0, 0}
	};

	exit_code         = 0;
	long_flag	  = 0;
	quiet_flag        = 0;
	verbosity         = 0;
	memset(&req_msg, 0, sizeof(shares_request_msg_t));
	slurm_init(NULL);
	if (priority_g_init() != SLURM_SUCCESS)
		fatal("failed to initialize priority plugin");
	log_init("sshare", opts, SYSLOG_FACILITY_DAEMON, NULL);

	while ((opt_char = getopt_long(argc, argv, "aA:ehlM:no:pPUu:vVm",
				       long_options, &option_index)) != -1) {
		switch (opt_char) {
		case (int)'?':
			fprintf(stderr, "Try \"sshare --help\" "
				"for more information\n");
			exit(1);
			break;
		case 'a':
			all_users = 1;
			break;
		case 'A':
			if (!req_msg.acct_list)
				req_msg.acct_list = list_create(xfree_ptr);
			slurm_addto_char_list(req_msg.acct_list, optarg);
			break;
		case 'e':
			_help_format_msg();
			exit(0);
			break;
		case 'h':
			print_fields_have_header = 0;
			break;
		case 'l':
			long_flag = 1;
			break;
		case 'M':
			FREE_NULL_LIST(clusters);
			if (slurm_get_cluster_info(&(clusters), optarg, 0)) {
				print_db_notok(optarg, 0);
				fatal("Could not get cluster information");
			}
			working_cluster_rec = list_peek(clusters);
			break;
		case 'm':
			options |= PRINT_PARTITIONS;
			break;
		case 'n':
			print_fields_have_header = 0;
			break;
		case 'o':
			xstrfmtcat(opt_field_list, "%s,", optarg);
			break;
		case 'p':
			print_fields_parsable_print =
			PRINT_FIELDS_PARSABLE_ENDING;
			break;
		case 'P':
			print_fields_parsable_print =
			PRINT_FIELDS_PARSABLE_NO_ENDING;
			break;
		case 'u':
			if (!xstrcmp(optarg, "-1")) {
				all_users = 1;
				break;
			}
			all_users = 0;
			if (!req_msg.user_list)
				req_msg.user_list = list_create(xfree_ptr);
			_addto_name_char_list(req_msg.user_list, optarg, 0);
			break;
		case 'U':
			options |= PRINT_USERS_ONLY;
			break;
		case 'v':
			quiet_flag = -1;
			verbosity++;
			break;
		case 'V':
			_print_version();
			exit(exit_code);
			break;
		case OPT_LONG_HELP:
		case OPT_LONG_USAGE:
			_usage();
			exit(0);
		case OPT_LONG_AUTOCOMP:
			suggest_completion(long_options, optarg);
			exit(0);
			break;
		case OPT_LONG_JSON:
			mimetype = MIME_TYPE_JSON;
			data_parser = optarg;
			serializer_required(MIME_TYPE_JSON);
			break;
		case OPT_LONG_YAML:
			mimetype = MIME_TYPE_YAML;
			data_parser = optarg;
			serializer_required(MIME_TYPE_YAML);
			break;
		default:
			exit_code = 1;
			fprintf(stderr, "getopt error, returned %c\n",
				opt_char);
			exit(exit_code);
		}
	}

	if (verbosity) {
		opts.stderr_level += verbosity;
		opts.prefix_level = 1;
		log_alter(opts, 0, NULL);
	}

	if (all_users) {
		if (req_msg.user_list
		   && list_count(req_msg.user_list)) {
			FREE_NULL_LIST(req_msg.user_list);
		}
		if (verbosity)
			fprintf(stderr, "Users requested:\n\t: all\n");
	} else if (verbosity && req_msg.user_list
	    && list_count(req_msg.user_list)) {
		fprintf(stderr, "Users requested:\n");
		list_itr_t *itr = list_iterator_create(req_msg.user_list);
		while ((temp = list_next(itr)))
			fprintf(stderr, "\t: %s\n", temp);
		list_iterator_destroy(itr);
	} else if (!req_msg.user_list || !list_count(req_msg.user_list)) {
		char *user = uid_to_string_or_null(getuid());
		if (user) {
			if (!req_msg.user_list) {
				req_msg.user_list = list_create(xfree_ptr);
			}
			list_append(req_msg.user_list, user);
			if (verbosity) {
				fprintf(stderr, "Users requested:\n");
				fprintf(stderr, "\t: %s\n", user);
			}
		}
	}

	if (verbosity && req_msg.acct_list && list_count(req_msg.acct_list)) {
		list_itr_t *itr = list_iterator_create(req_msg.acct_list);
		fprintf(stderr, "Accounts requested:\n");
		while ((temp = list_next(itr)))
			fprintf(stderr, "\t: %s\n", temp);
		list_iterator_destroy(itr);
	} else if (verbosity) {
		fprintf(stderr, "Accounts requested:\n\t: all\n");
	}

	if (clusters)
		exit_code = _multi_cluster(argc, argv, &req_msg);
	else
		exit_code = _single_cluster(argc, argv, &req_msg);

	FREE_NULL_LIST(req_msg.acct_list);
	FREE_NULL_LIST(req_msg.user_list);
	exit(exit_code);
}


static int _single_cluster(int argc, char **argv, shares_request_msg_t *req_msg)
{
	int rc = SLURM_SUCCESS;
	shares_response_msg_t *resp_msg = NULL;

	rc = slurm_associations_get_shares(req_msg, &resp_msg);
	if (rc) {
		slurm_perror("Couldn't get shares from controller");
		return rc;
	}

	if (mimetype) {
		DATA_DUMP_CLI_SINGLE(OPENAPI_SHARES_RESP, resp_msg, argc, argv,
				     NULL, mimetype, data_parser, rc);
	} else {
		process(resp_msg, options);
	}

	slurm_free_shares_response_msg(resp_msg);

	return rc;
}

static int _multi_cluster(int argc, char **argv, shares_request_msg_t *req_msg)
{
	list_itr_t *itr;
	bool first = true;
	int rc = 0, rc2;

	itr = list_iterator_create(clusters);
	while ((working_cluster_rec = list_next(itr))) {
		if (first)
			first = false;
		else
			printf("\n");
		printf("CLUSTER: %s\n", working_cluster_rec->name);
		rc2 = _single_cluster(argc, argv, req_msg);
		if (rc2)
			rc = 1;
	}
	list_iterator_destroy(itr);

	return rc;
}

/* returns number of objects added to list */
static int _addto_name_char_list_internal(list_t *char_list, char *name,
					  void *x)
{
	char *tmp_name = NULL;
	bool gid = *(bool *)x;

	if (isdigit(*name)) {
		uint32_t id = strtoul(name, NULL, 10);
		tmp_name = _convert_to_name(id, gid);
	} else
		tmp_name = xstrdup(name);

	if (!list_find_first(char_list, slurm_find_char_in_list, tmp_name)) {
		list_append(char_list, tmp_name);
		return 1;
	} else {
		xfree(tmp_name);
		return 0;
	}
}

/* returns number of objects added to list */
static int _addto_name_char_list(list_t *char_list, char *names, bool gid)
{
	if (!char_list) {
		error("No list was given to fill in");
		return 0;
	}

	return slurm_parse_char_list(char_list, names, &gid,
				     _addto_name_char_list_internal);
}

static char *_convert_to_name(uint32_t id, bool is_gid)
{
	char *name = NULL;

	if (is_gid) {
		if (!(name = gid_to_string_or_null(id))) {
			fprintf(stderr, "Invalid group id: %u\n", id);
			exit(1);
		}
	} else {
		if (!(name = uid_to_string_or_null(id))) {
			fprintf(stderr, "Invalid user id: %u\n", id);
			exit(1);
		}
	}
	return name;
}

static void _print_version(void)
{
	print_slurm_version();
	if (quiet_flag == -1) {
		long version = slurm_api_version();
		printf("slurm_api_version: %ld, %ld.%ld.%ld\n", version,
			SLURM_VERSION_MAJOR(version),
			SLURM_VERSION_MINOR(version),
			SLURM_VERSION_MICRO(version));
	}
}

/* _usage - show the valid sshare options */
void _usage(void){
	printf ("\
Usage:  sshare [OPTION]                                                    \n\
  Valid OPTIONs are:                                                       \n\
    -a or --all            list all users                                  \n\
    -A or --accounts=      display specific accounts (comma separated list)\n\
    -e or --helpformat     Print a list of fields that can be specified    \n\
                           with the '--format' option                      \n\
    --json[=data_parser]   Produce JSON output                             \n\
    -l or --long           include normalized usage in output              \n\
    -m or --partition      print the partition part of the association     \n\
    -M or --cluster=names  clusters to issue commands to.                  \n\
                           NOTE: SlurmDBD must be up.                      \n\
    -n or --noheader       omit header from output                         \n\
    -o or --format=        Comma separated list of fields (use             \n\
                           \"--helpformat\" for a list of available fields).\n\
    -p or --parsable       '|' delimited output with a trailing '|'        \n\
    -P or --parsable2      '|' delimited output without a trailing '|'     \n\
    -u or --users=         display specific users (comma separated list)   \n\
    -U or --Users          display only user information                   \n\
    -v or --verbose        display more information                        \n\
    -V or --version        display tool version number                     \n\
    --yaml[=data_parser]   Produce YAML output                             \n\
          --help           display this usage description                  \n\
          --usage          display this usage description                  \n\
                                                                           \n\n");
}

static void _help_format_msg(void)
{
	int i;

	for (i = 0; fields[i].name; i++) {
		if (i & 3)
			printf(" ");
		else if (i)
			printf("\n");
		printf("%-17s", fields[i].name);
	}
	printf("\n");
	return;
}
